lib/project.js

```diff
@@ -108,96 +145,172 @@ class Project {
+
+    return pkgs.map(
+      /** @type {PackageJSON|null} */ pkg => !!(pkg && pkg.type === 'module')
+    );
+  }
+
   /**
    * @param {string} basename
+   * @returns {Promise<{specifier: string, group: string, isModule: boolean}[]>}
+   * @private
    */
-  async loadInterfaceFiles(basename) {
-    const { dependencies = {}, devDependencies = {} } = this.packageJson;
-    const isProduction = process.env.NODE_ENV === 'production';
-    const deps = [
-      ...Object.keys(dependencies),
-      ...(isProduction ? [] : Object.keys(devDependencies)),
-    ];
-    debug('loadInterfaceFiles', { basename, isProduction });
+  async _getLocalTargets(basename) {
+    const extension = path.extname(basename);
+    const isModule = this.packageJson.type === 'module';
+    const patterns = `.{${getExtensions(isModule).join()}}`;
 
     const localFiles = await this.cachedGlob(
-      `{modules,lib}/*/${basename}.{js,mjs}`
+      `{modules,lib}/*/${basename}${!extension ? patterns : ''}`
     );
 
-    const localSpecifiers = localFiles.map(filename => {
+    return localFiles.map(filename => {
       const relativePath = filename.startsWith(`${this.root}/`)
         ? `.${filename.slice(this.root.length)}`
         : filename;
       const group = path.basename(path.dirname(filename));
-      return { specifier: relativePath, group };
+      return { specifier: relativePath, group, isModule };
     });
+  }
 
-    const depSpecifiers = deps.map(dep => {
+  /**
+   * @param {string} basename
+   * @param {string[]} deps
+   * @returns {Promise<{specifier: string, group: string, isModule: boolean}[]>}
+   * @private
+   */
+  async _getPackageTargets(basename, deps) {
+    const isModuleArr = await this.checkModules(deps);
+    return deps.map((dep, i) => {
       return {
         specifier: `${dep}/${basename}`,
         group: path.basename(dep),
+        isModule: isModuleArr[i],
       };
     });
+  }
+
+  /**
+   * @param {string} basename
+   * @returns {Promise<InterfaceFile[]>}
+   */
+  async loadInterfaceFiles(basename) {
+    const { dependencies = {}, devDependencies = {} } = this.packageJson;
+    const isProduction = process.env.NODE_ENV === 'production';
+    const deps = [
+      ...Object.keys(dependencies),
+      ...(isProduction ? [] : Object.keys(devDependencies)),
+    ];
+    debug('loadInterfaceFiles', { basename, isProduction });
+
+    const localSpecifiers = await this._getLocalTargets(basename);
+
+    const depSpecifiers = await this._getPackageTargets(basename, deps);
+
+    const allSpecifiers = [...localSpecifiers, ...depSpecifiers];
 
-    const results = await Promise.all(
-      [...localSpecifiers, ...depSpecifiers].map(
-        async ({ specifier, group }) => {
-          const moduleNamespace = /** @type {{ default: any }} */ (
-            await this.importOrNull(specifier)
-          );
-          if (moduleNamespace === null) return null;
+    /** @type {({default: *}|{default: string}|null|undefined)[]} */
```

Comment:
```suggestion
    /** @type {({ default: unknown } | null | undefined)[]} */
```

I think `unknown` is what you want in this case.  It's like `any` or `*`, but with the guarantee that you won't try to do anything with the contents.